### DISCUSSION TITLE
Treat `{ }` as a unit type

### DISCUSSION THREAD

**Discussion by hughesjs** [dotnet/csharplang#7383] [Language Ideas]
We've already got `{ }` in the sense of `not null` in pattern matching.

Why don't we extend this to be a catch-all unit type (type with no members, properties, fields, etc.).

For instance, if I had a generic `ApiResult<T>` class:

```cs
public class ApiResult<T> 
{
     T Payload { get; }
     Status Status { get; }
}
```

Having a unit type would let me return an `ApiResult` with no payload but that doesn't have to deal with the absolute mess that null causes just by doing `new ApiResult<{}>()`.

Now, I know I can (and have) just made a `public class UnitType {}` that does this just fine, but I'd like to see this be a first-class language feature.

**Comment by TahirAhmadov** [comment]
You can just do `new ApiResult<object?>()`.

Also, you can create a base class, `class ApiStatus { public Status Status { get; } }`, and return that.

**Reply by hughesjs** [reply]
I know you can, but 1) that's not the only application of unit types and 2) object? Is nullable, the whole point is to avoid nulls

**Reply by CyrusNajmabadi** [reply]
You could just use 'object' and have a Singleton value you return :-)

**Reply by HaloFour** [reply]
Or an empty `unit` struct, which I believe is what F# does here.

**Reply by khm1600** [reply]
The F# `unit` is an empty class and the value `()` is `null`.

**Comment by ufcpp** [comment]
0-tuple?

https://github.com/dotnet/csharplang/blob/64ff01fe49886d6f2edc5878a8f3e1cbd551f4a5/meetings/2016/LDM-2016-04-06.md?plain=1#L93

**Reply by hughesjs** [reply]
That might do the job tbf

**Comment by Rekkonnect** [comment]
Why not use `void` as the unit type? It already exists, it is a type, and C# actively forbids its usage. Also, this applies to existing methods that return `void`, and `Task` and `Task<void>` will become interchangeable. `void` should have never had this kind of treatment, despite being unusable. It's about the same as pointer types being actually unusable if not accessed via the `->` accessor. The only difference is pointer types not having any base classes, whereas `System.Void` is a `System.ValueType` ([for reference](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ7oGcCGAzAU3QG90ACCsgdTAEtgCAVATwAcCAhHLArACmBsCAezx9aUYACoAlDIDc5SjXpMhXHv0HtR4yQG0AunMVpK1OgxbsNvAUN0TgRoyaUUVV9dzvaRYp30AGmMFdwtVa04fLQcxOAAWNzNlSzUbGPsdeKQAdmTzT3TozSz/PgBlYDooAHMAYWEAW1YcOixhKALUyO9Sv10qmobm1rACMGSAX3R0OCQANjJEiK8M0qiyPxkScIA3NrIAYwgwcclNgF4toVNzAHcAC1oAGwIyPhOzgguhMlosGQoMJgECIC8XjsUhRSNDzMskABOT6nc7AKJhOHmL5oq7HVE/dFCAB0tiid0oMzh82RmKpQA)).

Thereby, by having `System.Void` as a usable type in C# by being the unit type, you can still invoke the methods `object` provides, just like a custom empty struct.

**Reply by HaloFour** [reply]
`void` already means something different, and the runtime forbids the use of `System.Void` in generics.  Void is different from unit, the former has no values while the latter has one value.

**Reply by Rekkonnect** [reply]
Void is technically a unit type, when you get a hold of one instance, its sole presence is its value.

Regardless, the real problem is that void already has runtime semantics, and not having both void and a separate unit type would be a large separator.

**Reply by HaloFour** [reply]
You can't get a value of `void`, that's the point.  There's no value to push/pop.  The stack must remain empty.  You can't conflate the two.

**Reply by 333fred** [reply]
I do think the runtime _could_ stitch this up to make it work, and if we were to make a unit type, I would want us to seriously investigate using `void`. It would be a large amount of work though, not just in allowing `void`, but also in making `Task` and `Task<void>` (and, by extension, many other such pairs of types) work as expected.

**Reply by HaloFour** [reply]
Possibly, but I'd argue that it'd make more sense to add a `System.Unit` type instead.  You'll never be able to fully make `void` mean the same as unit and the result would only be language warts where sometimes a value is required and sometimes you can't have a value at all.  Afterall, what would the be signature of `get_Result()` on `Task<void>`?  What about `List<void>.Add`?  You either have to synthesize a value to be used where no value can be used, or the runtime has pretend that nothing is on the stack.  `System.Unit` as a single-value struct just is infinitely easier all around.

**Reply by CyrusNajmabadi** [reply]
I'd really only want `void` to be the `unit` type if the runtime could *actually* unify things. in other words, unifying such that `Action<T>` was the same as `Function<T, void>`.  Otherwise, this will just leak through unpleasantly.  Having a new unit type avoids that.

**Reply by HaloFour** [reply]
Outside of the simplest cases I think that would be super weird, though.  Like, what is a `List<void>`?  How about `Action<T, void, T>`?  Is that the same as an `Action<T, T>` and a `Func<T, T, void>`?

**Reply by Rekkonnect** [reply]
> what is a `List<void>`?

Simply useless, but valid nonetheless.

> How about `Action<T, void, T>`? Is that the same as an `Action<T, T>` and a `Func<T, T, void>`?

Assuming delegates differentiate by type and not by signature, the types would be different. But the signatures would be identical.

Obligatory begging for delegates to have conversions between types enhanced to account for identical signatures.

**Reply by 333fred** [reply]
I really wouldn't want a blessed `System.Unit` type at all unless the runtime could unify things. And at that point, we should try and make it the actual unit type, `void`.

>  You'll never be able to fully make void mean the same as unit and the result would only be language warts where sometimes a value is required and sometimes you can't have a value at all.

I just don't agree with this statement. I do agree it would be a large amount of work, but we absolutely _could_ make it work.

**Reply by HaloFour** [reply]
> I really wouldn't want a blessed `System.Unit` type at all unless the runtime could unify things.

I think we're talking past each other.

Yes, if the runtime could go through the effort of supporting `void` in generics, such that `Func<T, void>` has the same signature as `Action<T>`, then sure, that's worth doing.

But, that's _not_ unit.  Unit is a different concept entirely, and one very easily supported today by the runtime.

**Reply by 333fred** [reply]
I don't agree that that's a different concept. If you want a unit type, then you want a type that represents the absence of a value. That's `void`. You want things like `List<void>` to collapse down to a count and nothing else, you want `Func<void>` to be equivalent to `Action`, etc.

Just because the runtime _could_ easily add `System.Unit` doesn't mean that it should, and I wouldn't want to put effort into a separate "this is a non-existent value" concept when we already have one.

**Reply by HaloFour** [reply]
> If you want a unit type, then you want a type that represents the absence of a value.

Unit has one value.  :)

**Reply by HaloFour** [reply]
Also, the runtime did add a blessed "unit" datatype, [ValueTuple](https://learn.microsoft.com/en-us/dotnet/api/system.valuetuple?view=net-7.0)

**Reply by 333fred** [reply]
> Unit has one value. :)

As a practical implementation detail, there may be an actual value. But that is not what unit _represents_, which is "I don't have a value".

**Reply by Rekkonnect** [reply]
Is ValueTuple treated in any special way? If not, it's just an empty struct that exists, just like Void.

**Reply by HaloFour** [reply]
> As a practical implementation detail, there may be an actual value. But that is not what unit represents, which is "I don't have a value".

Unit is a value, it just doesn't convey any information.

I do agree, though, that if the details of making `void` work as a generic type argument in the runtime can be worked out and implemented, that would likely scratch the vast majority of use cases where one might want a "unit" type.  I do have a ton of questions as to how that would work practically outside of the simplest of cases.
